README.rst
Moldy Bread
===========

.. image:: https://github.com/markpbaggett/moldybread/workflows/Build%20and%20Test/badge.svg
  :alt: GitHub Build and Test Badge
  :width: 25%

.. image:: https://github.com/markpbaggett/moldybread/workflows/Release%20and%20Publish/badge.svg
  :alt: GitHub Release and Publish Badge
  :width: 25%

An attempt to reimplement `whitebread <https://github.com/markpbaggett/whitebread/>`_ in `nim <https://nim-lang.org/>`_.

.. image:: https://markpbaggett.github.io/moldybread/moldy.gif
   :alt: Harvesting Metadata with No Pages with Moldy Bread
   

Installing and Building
-----------------------

**Installing the Release**

Moldy Bread is designed to be as portable as possible by being compiled in C.

The latest tagged release is available in `releases <https://github.com/markpbaggett/moldybread/releases>`_ and compiled with gcc. Depending on your architecture (X86 64Bit Linux),
you may be able to use the binary available here.

For security reasons, no authentication information is stored in the compiled version of the binary.

For this reason, you must point at a config.yml file.  For more information about this, `review the documentation <https://markpbaggett.github.io/moldybread/moldybread.html#defining-a-configdotyml>`_.

**Building the Application**

If you are unable to use the release binrary, you can build a binary with the files in this repository.

1. First, follow the instructions `here <https://nim-lang.org/install.html>`_ to install nim. Using choosenim is recommended.
2. Make sure you set your PATH appropriately so that nim and nimble can be found.
3. Make sure you have a c compiler like gcc or musl (if not, you'll get an error in step 6).
4. git clone git@github.com:markpbaggett/moldybread.git
5. cd moldybread
6. nimble install

Documentation
-------------

Documentation for all packages and public procs and types are generated automatically and hosted in this repository with `GitHub Pages <https://markpbaggett.github.io/moldybread/moldybread.html>`_ after each commit.

Look here for instructions on how to use all operations.


changelog.rst
======================
Moldy Bread Change Log
======================

0.1.6 - Not Yet Tagged for release
==================================

**New Features**:

* A file logger has been added based on Nim's default logging library.  For now, it logs all levels and is partially configurable within config.yml.

**Changes to FedoraRecord**:

* FedoraRecord now has 2 additional attributes: wait_time and retries.  These are there so you can declare a value for the number of times a method on FedoraRequest should be retried before in results in a `fatal` failure.  wait_time allows you to set the amount of milliseconds to wait before retrying.

0.1.5 - May 13, 2020
======================================

Tiny release specifically to allow for books to be exported and semantically organized.

**New Operations**:

* Download book pages and name semantically as book_pageNumber.extension.  See docs for more details.

**New Fedora Public Methods and Other Procs**:

* FedoraRequest.download_page_with_book_relatioship():  Downloads all matching datastreams related to a page with naming based on the book and page number.


0.1.4 - May 4, 2020
=============================

**Bug Fixes**:

* Address dc population bug where there was a 500 error if you had a dc string that included a space.

**New Operations**:

* Get a list of objects created or modified by a user with audit_responsibility.  See docs for more details.
* Find all objects matching your requested number of versions of a particular datastream.  See docs for more details.
* Find all objects that have no management restrictions (deny-management-functions).  See docs for more details.
* Find all exeptions to a specific XACML rule or action for all objects in a set.  See docs for more details.

**New Fedora Public Methods and Other Procs**:

* FedoraRequest.audit_responsibility(username): Returns a list of objects where a specific username has created or modified the object at some point.
* FedoraRequest.count_versions_of_datastreams(dsid): Returns pids with the total number of versions a specified datastream has.
* process_versions(): Requires a sequence of tuples with pids and number of versions are returns matches based on the specified operation and version value.

0.1.3 - January 10, 2020
=================================

**Bug Fixes**:

* Progress bars did not really work in production.  This was due to how the newProgressBar constructor was being called.  By default, the constructor sets the length of the bar to be 100, but for us this may be much bigger or much smaller. This has been addressed and should scale and tick appropriately.
* Update metadata did not respect whether or not a datastream was verisoned.  Because of this, any update on an unversioned datastream would switch the datastream back to being versionable.  This has been addressed.
* Update metadata was counting a successful update twice.  This has been addressed.
* Address thrown exception from empty result set.

**New Operations**:

* Download all versions of a datastream with download all versions.  See docs for more details.

**New Fedora Public Methods**:

* FedoraRequest.get_content_models(): This returns a sequence of tuples with the pid (as a string) and a human readable content model (as a string).

0.1.2 - January 5, 2020
=======================

**New Operations**:

* Find Unique Datastreams:  You can get a list of unique datastreams that belong to a result set. See docs for more details.
* Validate Checksums: You can validate checksums for a specific datastream or all datastreams in a result set.  See docs for more details.

**New Fedora Public Methods**:

* FedoraRequest.get_datastream():  This returns a sequence of tuples with the pid (as a string) and a sequence of datastreams (as TaintedStrings) that belong to it. You can specify whether you want the entire datastream profile or just the datastreams (as HTML only) and specify a date that you want to base the request on. While this method is public, it is not currently called from the executable.
* FedoraRequest.validate_checksums(): See new operations for more details.
* FedoraRequest.find_distinct_datastreams(): See new operations for more details.

**Other**:

* Yaml config flag is now optional if you have a config/config.yml in the directory you call your executable from.
* FedoraRecord.get() now returns the response body of an HTTP request as a string or an empty string rather than true or false. This is mostly to help FedoraRequest.get_datastream().
* xmlhelper and parse_data()* has been added.

0.1.1 - January 1, 2020
=======================

**Documentation**:

* Build and release download instructions have been added.

**Operations**:

* Harvest Metadata and Harvest Metadata No Pages were renamed to Harvest Datastream to make things more agnostic. MODS is still default.
* Update Metadata and Delete Old Versions was added as a separate operation to optionally allow old versions of datastreams to be deleted on update.
* Change Object State allows you to batch change the state of all objects in a set to Active, Inactive, or Deleted.
* Find Objects Missing a Datastream was added to check for objects in a result set missing a particular datastream.
* Download Datastream Histories as XML was added to allow you to get a log of all events occuring on a datastream over time.
* Download Datastream at Date allows you to get a datastream at a specific date and time.
* For more information about any of this, read the docs.

**Other Notable Changes**:

* More extensions were added to cover expected mime types for datastreams.
* FedoraRecord.get() was renamed to FedoraRecord.download().  Download has the side affect of serializing to disk. A new get was created for returning values to other operations.

0.1.0 - December 17, 2020
=========================

* Initial release.

tests/test1.nim
import unittest, xmltools, moldybread, moldybreadpkg/fedora, typetraits, moldybreadpkg/xmlhelper, moldybreadpkg/rdfhelper

suite "Test Public Types Initialization":
  echo "Test Public Types Initialization"

  setup:
    let fedora_connection = initFedoraRequest(
      pid_part="test",
      dc_values="title:Pencil;contributor:Wiley",
      auth=("admin", "password"),
      url="http://localhost",
      max_results=20,
      output_directory="/home/user/output")

  test "FedoraRequest Initialization":
    check(fedora_connection.base_url == "http://localhost")
    check(fedora_connection.max_results == 20)

suite "Test Fedora Connection Methods":
  echo "Fedora Connection Methods"
  
  setup:
    let fedora_connection = initFedoraRequest(pid_part="garbagenamespace")
  
  test "Population Works as Expected":
    doAssert(typeof(fedora_connection.populate_results()) is seq[string])

  test "Harvest Metadata":
    doAssert(typeof(fedora_connection.harvest_datastream("DC")) is Message)

  test "Harvest Metadata No Pages":
    doAssert(typeof(fedora_connection.harvest_datastream_no_pages("DC")) is Message)

suite "Parse Data XML Helper":
  echo "Parse Data XML Helper Tests"

  setup:
    let
      some_xml = """<?xml version="1.0" encoding="UTF-8"?><datastreamProfile  xmlns="http://www.fedora.info/definitions/1/0/management/"  xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.fedora.info/definitions/1/0/management/ http://www.fedora.info/definitions/1/0/datastreamProfile.xsd" pid="test:9" dsID="MODS"><dsLabel>MODS Record</dsLabel>
      <dsVersionID>MODS.5</dsVersionID>
      <dsCreateDate>2019-12-19T02:50:24.322Z</dsCreateDate>
      <dsState>A</dsState>
      <dsMIME>application/xml</dsMIME>
      <dsFormatURI></dsFormatURI>
      <dsControlGroup>X</dsControlGroup>
      <dsSize>178</dsSize>
      <dsVersionable>true</dsVersionable>
      <dsInfoType></dsInfoType>
      <dsLocation>test:9+MODS+MODS.5</dsLocation>
      <dsLocationType></dsLocationType>
      <dsChecksumType>SHA-1</dsChecksumType>
      <dsChecksum>f2e60f8860158d6d175bdd3c2710928c79a5d024</dsChecksum>
      <dsChecksumValid>true</dsChecksumValid>
      </datastreamProfile>
      """
      an_element = "dsChecksumValid"

  test "parse_data works as expected":
    assert parse_data(some_xml, an_element) == @["true"]

suite "Get Attribute of Elements Test":
  echo "Get Attribute of Elements"

  setup:
    let
      some_xml = """<rdf:RDF xmlns:fedora="info:fedora/fedora-system:def/relations-external#" xmlns:fedora-model="info:fedora/fedora-system:def/model#" xmlns:islandora="http://islandora.ca/ontology/relsext#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
      <rdf:Description rdf:about="info:fedora/test:6">
        <islandora:isPageOf rdf:resource="info:fedora/test:3"></islandora:isPageOf>
        <islandora:isSequenceNumber>2</islandora:isSequenceNumber>
        <islandora:isPageNumber>2</islandora:isPageNumber>
        <islandora:isSection>1</islandora:isSection>
        <fedora:isMemberOf rdf:resource="info:fedora/test:3"></fedora:isMemberOf>
        <fedora-model:hasModel rdf:resource="info:fedora/islandora:pageCModel"></fedora-model:hasModel>
        <islandora:generate_ocr>TRUE</islandora:generate_ocr>
        </rdf:Description>
        </rdf:RDF>"""
      an_element = "fedora-model:hasModel"
      an_attribute = "rdf:resource"

  test "get_attribute_of_element works as expected":
    assert get_attribute_of_element(some_xml, an_element, an_attribute) == @["info:fedora/islandora:pageCModel"]

suite "Get Text of an Element Based on Attribute and Value of Attribute":
  echo "Get Text of an Element Based on Attribute and Value of Attribute"

  setup:
    let
      some_xml = """<mods xmlns="http://www.loc.gov/mods/v3" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/mods/v3 http://www.loc.gov/standards/mods/v3/mods-3-5.xsd"><identifier type="filename">yoyoyo.yo</identifier>
        <identifier type="local">lady-vols-basketball_2010</identifier>
        <titleInfo supplied="yes">
        <title>Tennessee Lady Volunteers basketball media guide, 2010-2011</title>
        </titleInfo>
        <titleInfo type="alternative" displayLabel="Cover Title">
          <title>2010-11 Lady Vol basketball: victory lane</title>
        </titleInfo>
        <originInfo>
          <dateIssued>2010</dateIssued>
          <dateIssued encoding="edtf" keyDate="yes">2010</dateIssued>
          <publisher>University of Tennessee, Knoxville. Department of Athletics</publisher>
          <place>
            <placeTerm valueURI="http://id.loc.gov/authorities/names/n79109786">Knoxville (Tenn.)</placeTerm>
          </place>
        </originInfo>
        <abstract>Lady Volunteers basketball program from 2010.</abstract>
        <physicalDescription>
          <form authority="aat" valueURI="http://vocab.getty.edu/aat/300311670">booklets</form>
          <extent>212 pages</extent>
        </physicalDescription>
        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85012111"><topic>Basketball</topic></subject>
        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh2004010434"><topic>College sports for women</topic></subject>
        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85028338"><topic>College sports</topic></subject>
        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85012123"><topic>Basketball players</topic></subject>
        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85147485"><topic>Women basketball players</topic></subject>
        <subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n80003889"><name><namePart>University of Tennessee, Knoxville</namePart></name></subject><subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n88072771">
        <name><namePart>Summitt, Pat Head, 1952-2016</namePart></name></subject><subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n88072776"><name><namePart>Lady Volunteers (Basketball team)</namePart></name></subject>
        <subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n79109786">
        <geographic>Knoxville (Tenn.)</geographic>
        <cartographics>
          <coordinates>35.96064, -83.92074</coordinates>
        </cartographics>
        </subject>
        <typeOfResource>text</typeOfResource>
        <classification authority="lcc">LD5296.A7</classification>
        <relatedItem displayLabel="Project" type="host"><titleInfo><title>University of Tennessee Lady Volunteers Basketball Media Guides</title></titleInfo></relatedItem>
        <location>
        <physicalLocation valueURI="http://id.loc.gov/authorities/names/no2014027633">University of Tennessee, Knoxville. Special Collections</physicalLocation>
        </location>
        <recordInfo>
        <recordContentSource valueURI="http://id.loc.gov/authorities/names/n87808088">University of Tennessee, Knoxville. Libraries</recordContentSource>
        </recordInfo>
        <accessCondition type="use and reproduction" xlink:href="http://rightsstatements.org/vocab/InC/1.0/">In Copyright</accessCondition>
        </mods>"""
      an_element = "identifier"
      attribute_and_value = ("type", "local")
  
  test "Test that we can get value of identifer with type=local":
    assert get_text_of_element_with_attribute(some_xml, an_element, attribute_and_value) == @["lady-vols-basketball_2010"]

suite "Test RDF Helper":
  echo "Test RDF Helper"

  setup:
    let
      turtle = "<info:fedora/test:22> <info:fedora/fedora-system:def/relations-external#isMemberOf> <info:fedora/test:21> ."

  test "newTriple contructor works for subjects as expected":
    assert newTriple(turtle).subject == "<info:fedora/test:22>"
  
  test "newTriple constructor works for predicates as expected":
    assert newTriple(turtle).predicate == "<info:fedora/fedora-system:def/relations-external#isMemberOf>"


src/moldy.nim
import httpClient, streams, strutils, xmltools

type
  FedoraConnection* = ref object
    base_url*: string
    results*: seq[string]
    query*: string
    max_results*: int

var client = newHttpClient()

proc grab_pids(response: string): seq[string] =
  var pids: seq[string] = @[]
  let xml_response = Node.fromStringE(response)
  let results = $(xml_response // "pid")
  for word in split(results, '<'):
    let new_word = word.replace("/", "").replace("pid>", "")
    if len(new_word) > 0:
      pids.add(new_word)
  return pids

proc get_token(response: string): string =
  let xml_response = Node.fromStringE(response)
  let results = $(xml_response // "token")
  var token: string = ""
  if results.len > 0:
    token = results.replace("<token>", "").replace("</token>", "")
  return token

proc populate_results(connection: FedoraConnection): seq[string] =
  var pids: seq[string] = @[]
  var new_pids: seq[string] = @[]
  var token: string = "temporary"
  var url: string = connection.base_url & "/fedora/objects?query=pid%7E" & connection.query & "*&pid=true&resultFormat=xml&maxResults=" & $connection.max_results
  var response: string = ""
  while token.len > 0:
    response = client.getContent(url)
    new_pids = grab_pids(response)
    for pid in new_pids:
      pids.add(pid)
    token = get_token(response)
    url = connection.base_url & "/fedora/objects?query=pid%7E" & connection.query & "*&pid=true&resultFormat=xml&maxResults=" & $connection.max_results & "&sessionToken=" & token
  return pids  

var fedora_connection: FedoraConnection = FedoraConnection(base_url:"http://localhost:8080", query: "test", max_results: 2)
fedora_connection.results = populate_results(fedora_connection)
echo fedora_connection.results


src/moldybreadpkg/xacml.nim
import xmltools, strutils, strformat

type
  XACMLRule* = ref object of RootObj
    rule_id*: string
    effect*: string
    target_actions*: seq[string]
    excepted_roles_and_logins*: seq[string]
    body*: string

proc get_node(xml, node: string): seq[string] =
  let
    xml_response = Node.fromStringE(xml)
    new_data = $(xml_response // node)
  for part in new_data.split(fmt"<{node}"):
    if part != "":
      result.add(fmt"<{node}{part}")

proc build_actions(xml: string): seq[string]=
  let
    action_list = get_node(xml, "ActionMatch")
  for action in action_list:
    let
      attribute_values = get_node(action, "AttributeValue")
    for value in attribute_values:
      result.add(value.split('>')[1].split('<')[0])

proc get_attribute_value(xml, target_attribute: string): string =
  for value in xml.split(" "):
    if value.contains(target_attribute):
      return value.split("=")[1].replace("\"", "")

proc build_logins_and_roles(xml: string): seq[string] =
  let
    conditions_list = get_node(xml, "Condition")
  for condition in conditions_list:
    let
      attribute_values = get_node(condition, "AttributeValue")
    for value in attribute_values:
      result.add(value.split('>')[1].replace("</AttributeValue", ""))

proc newXACMLRule(rule: string): XACMLRule =
  XACMLRule(
    body: rule,
    rule_id: get_attribute_value(rule, "RuleId").replace("\n", ""),
    effect: get_attribute_value(rule, "Effect").replace(">", "").replace("<Target", "").replace("\n", ""),
    target_actions: build_actions(rule),
    excepted_roles_and_logins: build_logins_and_roles(rule)
  )

proc parse_rules*(response: string): seq[XACMLRule] =
  ## Parses a XACML policy as a string and returns XACMLRules with information about the policy.
  ##
  ## To be clear, this isn't a true interpretation of XACML 1.0, but should work with our use cases.
  ##
  let
    xml_response = Node.fromStringE(response)
    rules = $(xml_response // "Rule")
  for rule in rules.split("<Rule"):
    if rule != "":
      result.add(newXACMLRule(fmt"<Rule{rule}"))


src/moldybreadpkg/preserve.nim
import httpclient, base64, strformat, xmlhelper, strutils, yaml/serialization, streams, os

type
  PreservationObject* = ref object
    ## Handles Preservation Object
    pid*, content_model, parent, base_uri*, external_relationships: string
    dsid_and_checksum: seq[(string, string)]
    client*: HttpClient
  
  ContentModel = object
    dsids: seq[string]

  PreservationSettings = object
    ## Type to represent preservation settings
    book, page, large_image, basic_image: ContentModel
  
  AccessError* = object of Exception

  LookupError* = object of Exception

proc read_yaml_config(file_path: string): PreservationSettings =
  var file_stream = newFileStream(file_path)
  load(file_stream, result)
  file_stream.close()

proc content_model_translator(content_model: string): string = 
  case content_model
  of "islandora:bookCModel":
    "book"
  of "islandora:pageCModel":
    "page"
  of "islandora:sp_large_image_cmodel":
    "large_image"
  of "islandora:sp_basic_image":
    "basic_image"
  else:
    raise newException(LookupError, fmt"{content_model} could not be translated.")

proc get_extension(header: HttpHeaders): string =
  case $header["content-type"]
  of "application/xml", "text/xml", "application/rdf+xml":
    ".xml"
  of "text/plain":
    ".txt"
  of "text/html":
    ".html"
  of "application/pdf":
    ".pdf"
  of "image/tiff":
    ".tif"
  of "image/jp2":
    ".jp2"
  of "image/jpeg":
    ".jpg"
  else:
    ".bin"

proc serialize_file(filename, contents, output_directory: string) =
  if not existsDir(output_directory):
    createDir(output_directory)
  let path = fmt"{output_directory}/{filename}"
  writeFile(path, contents)

proc initPreservationObject*(pid, uri: string, auth: (string, string)): PreservationObject =
  let client = newHttpClient()
  client.headers["Authorization"] = "Basic " & base64.encode(auth[0] & ":" & auth[1])
  PreservationObject(client: client, base_uri: uri, pid: pid)

method get_external_relationships*(this: PreservationObject): string {. base .} =
  let
    request = this.client.request(fmt"{this.base_uri}/fedora/objects/{this.pid}/datastreams/RELS-EXT/content", httpMethod = HttpGet)
  if request.status == "200 OK":
    request.body
  else:
    raise newException(AccessError, fmt"The request for the content model of {this.pid} failed with {request.status}.")

method get_checksum(this: PreservationObject, dsid: string): string {. base .} =
  let
    request = this.client.request(fmt"{this.base_uri}/fedora/objects/{this.pid}/datastreams/{dsid}?validateChecksum=false&format=xml", httpMethod = HttpGet)
  if request.status == "200 OK":
    parse_data(request.body, "dsChecksum")[0]
  else:
    raise newException(AccessError, fmt"The request for the {dsid} datastream on {this.pid} failed with {request.status}.")

method get_preservation_datastreams(this: PreservationObject, settings: PreservationSettings): seq[string] {. base .} =
  case this.content_model
  of "book":
    settings.book.dsids
  of "page":
    settings.page.dsids
  of "large_image":
    settings.large_image.dsids
  of "basic_image":
    settings.basic_image.dsids
  else:
    raise newException(LookupError, fmt"Unable to retrieve relevant preservation datastreams for {this.content_model}.")

method get_content_model(this: PreservationObject): string {. base .} =
  content_model_translator(get_attribute_of_element(this.external_relationships, "fedora-model:hasModel", "rdf:resource")[0].replace("info:fedora/", ""))

method get_parent_and_page_number(this: PreservationObject): (string, string) {. base .} =
  (
    get_attribute_of_element(this.external_relationships, "islandora:isPageOf", "rdf:resource")[0].replace("info:fedora/", ""),
    parse_data(this.external_relationships, "islandora:isPageNumber")[0]
  )
  
method download(this: PreservationObject, datastream: string): string {. base .} =
  let request = this.client.request(fmt"{this.base_uri}/fedora/objects/{this.pid}/datastreams/{datastream}/content")
  if request.status == "200 OK":
    if this.content_model == "page":
      let rels = this.get_parent_and_page_number()
      serialize_file(fmt"{rels[0]}-page{rels[1]}{get_extension(request.headers)}", request.body, "/home/mark/nim_projects/moldybread/experiment")
      fmt"{rels[0]}-page{rels[1]}{get_extension(request.headers)}"
    else:
      serialize_file(fmt"{this.pid}{get_extension(request.headers)}", request.body, "/home/mark/nim_projects/moldybread/experiment")
      fmt"{this.pid}{get_extension(request.headers)}"
  else:
    raise newException(AccessError, fmt"Could not download {this.pid}'s {datastream} datastream.")

method preserve*(this: PreservationObject): seq[(string, string)] {. base .} =
  let preservation_settings = read_yaml_config(fmt"{getCurrentDir()}/config/preserve.yml")
  this.external_relationships = this.get_external_relationships()
  this.content_model = this.get_content_model()
  for datastream in this.get_preservation_datastreams(preservation_settings):
    let filename = this.download(datastream)
    this.dsid_and_checksum.add((filename, this.get_checksum(datastream)))
  this.dsid_and_checksum


src/moldybreadpkg/xmlhelper.nim
import xmltools, strutils, strformat

proc parse_data*(response, element: string): seq[string] =
  ## Takes XML as a string and an element name and returns the text nodes of each element as a sequence of strings.
  ##
  ## Example:
  ##
  ## .. code-block:: nim
  ##
  ##    let
  ##          some_xml = """<?xml version="1.0" encoding="UTF-8"?><datastreamProfile  xmlns="http://www.fedora.info/definitions/1/0/management/"  xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.fedora.info/definitions/1/0/management/ http://www.fedora.info/definitions/1/0/datastreamProfile.xsd" pid="test:9" dsID="MODS"><dsLabel>MODS Record</dsLabel>
  ##          <dsVersionID>MODS.5</dsVersionID>
  ##          <dsCreateDate>2019-12-19T02:50:24.322Z</dsCreateDate>
  ##          <dsState>A</dsState>
  ##          <dsMIME>application/xml</dsMIME>
  ##          <dsFormatURI></dsFormatURI>
  ##          <dsControlGroup>X</dsControlGroup>
  ##          <dsSize>178</dsSize>
  ##          <dsVersionable>true</dsVersionable>
  ##          <dsInfoType></dsInfoType>
  ##          <dsLocation>test:9+MODS+MODS.5</dsLocation>
  ##          <dsLocationType></dsLocationType>
  ##          <dsChecksumType>SHA-1</dsChecksumType>
  ##          <dsChecksum>f2e60f8860158d6d175bdd3c2710928c79a5d024</dsChecksum>
  ##          <dsChecksumValid>true</dsChecksumValid>
  ##          </datastreamProfile>
  ##          """
  ##      an_element = "dsChecksumValid"
  ##    assert parsedata(some_xml, an_element) == @["true"]
  ##
  let
    xml_response = Node.fromStringE(response)
    results = $(xml_response // element)
  for node in split(results, '<'):
    let value = node.replace("/", "").replace(fmt"{element}>", "")
    if len(value) > 0:
      result.add(value)

proc get_attribute_of_element*(response, element, attribute: string): seq[string] =
  ## Terrible code to try to find matching attribute values for nodes in a string of XML.
  ##
  ## Example:
  ##
  ## .. code-block:: nim
  ##
  ##     let
  ##       some_xml = """<rdf:RDF xmlns:fedora="info:fedora/fedora-system:def/relations-external#" xmlns:fedora-model="info:fedora/fedora-system:def/model#" xmlns:islandora="http://islandora.ca/ontology/relsext#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
  ##         <rdf:Description rdf:about="info:fedora/test:6">
  ##         <islandora:isPageOf rdf:resource="info:fedora/test:3"></islandora:isPageOf>
  ##         <islandora:isSequenceNumber>2</islandora:isSequenceNumber>
  ##         <islandora:isPageNumber>2</islandora:isPageNumber>
  ##         <islandora:isSection>1</islandora:isSection>
  ##         <fedora:isMemberOf rdf:resource="info:fedora/test:3"></fedora:isMemberOf>
  ##         <fedora-model:hasModel rdf:resource="info:fedora/islandora:pageCModel"></fedora-model:hasModel>
  ##         <islandora:generate_ocr>TRUE</islandora:generate_ocr>
  ##         </rdf:Description>
  ##         </rdf:RDF>"""
  ##       an_element = "fedora-model:hasModel"
  ##       an_attribute = "rdf:resource"
  ##     assert get_attribute_of_element(some_xml, an_element, an_attribute) == @["info:fedora/islandora:pageCModel"]
  let
    xml_response = Node.fromStringE(response)
    results = $(xml_response // element)
  for node in split(results, '<'):
    if attribute in node:
      result.add(node.split(attribute)[1].split("\"")[1])

proc get_text_of_element_with_attribute*(response, element: string, attribute_with_value: (string, string)): seq[string] =
  ## Grabs text values of elements with associated attributes.  Use with caution.
  ##
  ## Example:
  ##
  ## .. code-block:: nim
  ##
  ##    let
  ##      some_xml = """<mods xmlns="http://www.loc.gov/mods/v3" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/mods/v3 http://www.loc.gov/standards/mods/v3/mods-3-5.xsd">
  ##        <identifier type="local">lady-vols-basketball_2010</identifier>
  ##        <titleInfo supplied="yes"><title>Tennessee Lady Volunteers basketball media guide, 2010-2011</title></titleInfo>
  ##        <titleInfo type="alternative" displayLabel="Cover Title"><title>2010-11 Lady Vol basketball: victory lane</title></titleInfo>
  ##        <originInfo><dateIssued>2010</dateIssued><dateIssued encoding="edtf" keyDate="yes">2010</dateIssued>
  ##        <publisher>University of Tennessee, Knoxville. Department of Athletics</publisher>
  ##        <place>
  ##        <placeTerm valueURI="http://id.loc.gov/authorities/names/n79109786">Knoxville (Tenn.)</placeTerm>
  ##        </place>
  ##        </originInfo>
  ##        <abstract>Lady Volunteers basketball program from 2010.</abstract>
  ##        <physicalDescription>
  ##        <form authority="aat" valueURI="http://vocab.getty.edu/aat/300311670">booklets</form>
  ##        <extent>212 pages</extent>
  ##        </physicalDescription>
  ##        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85012111">
  ##        <topic>Basketball</topic>
  ##        </subject>
  ##        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh2004010434">
  ##        <topic>College sports for women</topic>
  ##        </subject>
  ##        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85028338">
  ##        <topic>College sports</topic>
  ##        </subject>
  ##        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85012123">
  ##        <topic>Basketball players</topic>
  ##        </subject>
  ##        <subject authority="lcsh" valueURI="http://id.loc.gov/authorities/subjects/sh85147485">
  ##        <topic>Women basketball players</topic>
  ##        </subject>
  ##        <subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n80003889">
  ##        <name>
  ##        <namePart>University of Tennessee, Knoxville</namePart>
  ##        </name></subject>
  ##        <subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n88072771"><name><namePart>Summitt, Pat Head, 1952-2016</namePart></name></subject>
  ##        <subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n88072776"><name><namePart>Lady Volunteers (Basketball team)</namePart></name></subject>
  ##        <subject authority="naf" valueURI="http://id.loc.gov/authorities/names/n79109786"><geographic>Knoxville (Tenn.)</geographic><cartographics><coordinates>35.96064, -83.92074</coordinates></cartographics></subject>
  ##        <typeOfResource>text</typeOfResource>
  ##        <classification authority="lcc">LD5296.A7</classification>
  ##        <relatedItem displayLabel="Project" type="host"><titleInfo><title>University of Tennessee Lady Volunteers Basketball Media Guides</title></titleInfo></relatedItem>
  ##        <location><physicalLocation valueURI="http://id.loc.gov/authorities/names/no2014027633">University of Tennessee, Knoxville. Special Collections</physicalLocation></location>
  ##        <recordInfo><recordContentSource valueURI="http://id.loc.gov/authorities/names/n87808088">University of Tennessee, Knoxville. Libraries</recordContentSource></recordInfo>
  ##        <accessCondition type="use and reproduction" xlink:href="http://rightsstatements.org/vocab/InC/1.0/">In Copyright</accessCondition>
  ##        </mods>"""
  ##     element = "identifier"
  ##     attribute_with_value = ("type", "local")
  ##     assert get_text_of_element_with_attribute(some_xml, element, attribute_with_value) == @["lady-vols-basketball_2010"]
  ##
  let
    xml_response = Node.fromStringE(response)
    results = $(xml_response // element)
  for node in split(results, '<'):
    if fmt"""{attribute_with_value[0]}="{attribute_with_value[1]}"""" in node:
      result.add(node.split('>')[1])
